### PR TITLE
[dsx] new implementation 

### PR DESCRIPTION
**Pull request by npomfret**

### CODE REVIEW

**Comment by npomfret** [comment]
... any help would be appreciated!

**Comment by npomfret** [comment]
@kroitor ok, I think the initial implementation is ready - there's more to do at some point as I've only covered about 1/2 the api

**Comment by npomfret** [comment]
Is the build dead?

**Comment by kroitor** [comment]
@npomfret travis CI is having trouble with the infrastructure recently, but it will unstuck eventually or i will restart it as soon as their system becomes available

**Comment by npomfret** [comment]
@kroitor thanks - at least it's green now

**Comment by kroitor** [comment]
@npomfret i'll try to get to it before the end of next week

**Comment by npomfret** [comment]
@kroitor could we get this merged?

**Comment by kroitor** [comment]
@npomfret yep, if you can help resolve the conflict – that would speed things up, i'll get to it shortly.

File: js/dsx.js [hunk 1]
```diff
@@ -201,532 +250,262 @@ module.exports = class dsx extends Exchange {
...
-            this.safeFloat (ohlcv, 'close'),
-            this.safeFloat (ohlcv, 'amount'),
-        ];
+    async fetchClosedOrders (symbol = undefined, since = undefined, limit = undefined, params = {}) {
+        await this.fetchOrders (symbol, since, limit, params);
+        const orders = this.filterBy (this.orders, 'status', 'closed');
+        return this.filterBySymbolSinceLimit (orders, symbol, since, limit);
     }
 
-    async fetchOHLCV (symbol, timeframe = '1m', since = undefined, limit = undefined, params = {}) {
```
Comment (kroitor) [inline js/dsx.js:590]: @npomfret We could save this method by a replacing `pair` with `instrument` in it. What do you say?

Comment (kroitor) [reply js/dsx.js:590]: I'd suggest to use url param names as they're named in their API docs – for less confusion with params overrides.

Comment (npomfret) [reply js/dsx.js:590]: Does the new API have that feature?

Comment (npomfret) [reply js/dsx.js:590]: https://api.dsxglobal.com/#candles ?

Comment (npomfret) [reply js/dsx.js:590]: I think the old API is completely dead anyway, so merging and adding this method later is an option.

Comment (npomfret) [reply js/dsx.js:590]: Have added fetchOHLCV.  Looks like DSX is VERY similar, or possibly the same as hitbtc.  I would be nice to get some code reuse, but I don't think inheritance (`dsx extends hitbtc`) is the answer though, it inevitably leads to as many problems as it solves.  Perhaps if they both inherited from a common super class...

`class hitbtc extends AbstractHitbtcAndDsxFunctionality {...`

`class dsx extends AbstractHitbtcAndDsxFunctionality {...`

Or better (IMO), could we move the common functions out and have each implementation delegate by cherry picking which functions it wants to use?

**Comment by kroitor** [comment]
> I don't really understand why there's a conflict here

It's because the same place in the file has been edited and reverted back in master, while the file in PR is forked from an earlier master version. You can help resolve it, if you merge the master branch into that fork – the conflicts will be outlined upon the merge, so you could remove the redundant conflicting lines. Alternatively, you can simply submit a new PR (a new fork), overwriting the dsx.js with your new version – there will be no conflict, because that new fork will be cloned from the current master. If that is something you can do – that would really help. If not – I'll do that myself as soon as i can.

**Comment by npomfret** [comment]
I have merged, that what I don't understand

**Comment by npomfret** [comment]
Sorry, didn't push my last merge

**Comment by kroitor** [comment]
@npomfret thx so much! I'll get to merging it within a few hours.

**Comment by npomfret** [comment]
@kroitor thanks for merging

Could we discuss dong `class dsx extends hitbtc` ?  I really think inheritance like this causes as many problems as it solves, more actually.

**Comment by kroitor** [comment]
> Could we discuss class dsx extends hitbtc ?

Sure

> I really think inheritance like this causes as many problems as it solves, more actually.

To my mind their API docs look almost identical now, so there's not too much sense in doing the duplication, however, if we find significant differences – we could simply clone the hitbtc's implementation and edit the differing methods. Does that make sense? Do you have any specific issues with it right now?

**Comment by npomfret** [comment]
I've just spent so much in ccxt where inheritance has been used trying to figure out which class does what.  We don't have a consistent approach for dealing with subtle differences.  Exchanges that share some behaviour at some point inevitably seem to move apart as time goes on, but in ccxt we don't have a clear or consistent way of dealing with it and the code ends up being very difficult to follow. `sign` methods seem to become particularly messy.  And, by extending one class form the other, we seem to almost invite _hackery_ when it comes to implementing these small differences.

If you are really keen in using inheritance, I'd strongly suggest having both hitbtc and dsx extend a common superclass, rather than what we have now.  And, if the two exchanges start to move apart in some way, we could remove the method(s) in question from the common superclass and put it/them into the underlying implementations.  

I'd prefer not to use inheritance though if there were an alternative.  Normally I'd go for delegation.  We could have a bunch of functions defined somewhere and each implementation can cherry pick which ones they'd want to use.  Not sure how cross complication would cope, so perhaps this is a step too far for now.

In this case, how sure are we that both exchanges support the paths that are now in the hitbtc implementation?  I see you added a lot in there.

**Comment by kroitor** [comment]
> In this case, how sure are we that both exchanges support the paths that are now in the hitbtc implementation?

I would suggest to test it. Unfortunately, since they migrated to a new domain, they've dropped my account and i don't have an API key on the new account to test it on my side right now, because they require KYC, which takes time. If you could help test it and reveal the differences – i'd be happy to move away from inheritance in this case. If they turn out to be identical now – i don't see a reason to duplicate 100% of the code.

In short, I agree with most of your points if the APIs have significant differences. I don't insist on inheritance at all. Just thought that in this particular case it could save us some time reinventing the wheel again. If it introduces more problems than it solves – we shouldn't extend from hitbtc, and should simply denormalize (copy-paste) the existing implementations, where they fit.

**Comment by npomfret** [comment]
The problem is that if you start with inheritance it tends to stick around even when it's not the right thing to do.  And that results in hack after hack.  I've seen it over and over again, not just in this project, but every project I've ever worked on.  

My suggestion of a common superclass would result in no duplication, but it would make it easier to differentiate between the exchanges when the time comes. If we don't do it now, it won't get done and the first difference will get hacked in - I guarantee it.

Can I suggest we take this approach now?  And can I also suggest we move the paths back out if we aren't sure that hitbtc supports all the same paths that dsx does?   Every path I added to the new dsx implemantion is definitely supported by dsx, but I can say the same for hitbtc.  You see, by using the fake _convenience_ of inheritance we've already introduced a problem here (by combining all these paths together).

**Comment by npomfret** [comment]
@kroitor can we pease revert your changes?  I'm getting a lot of errors

**Comment by kroitor** [comment]
> I guarantee it.

I don't doubt that – I agree with you on that. But the question is: are there differences now?

> My suggestion of a common superclass would result in no duplication, but it would make it easier to differentiate between the exchanges when the time comes.

I would suggest to duplicate the implementation then. Otherwise it will be much harder to differentiate between exchanges and "superclasses" in the long run, and it will require a lot of changes in the build structure, which aren't really necessary if we just duplicate the code. I would refrain from making superclasses due to its own complexities.

However, I still think that we should first test it for the purpose of certainty. If you insist on splitting them without testing – we can do that as well, though, I would still test them to find out the differences.

> can we pease revert your changes? I'm getting a lot of errors

Can you point them out? I will do the revert based on that.

**Comment by npomfret** [comment]
I've created tickets for the errors I've spotted so far

**Comment by npomfret** [comment]
I'm not sure I understand why we can have a common superclass.  Will something like this not work?

    class dsx extends AbstractDsxAndHitbtcFunctionality {...

**Comment by kroitor** [comment]
> Will something like this not work?

Nope, because a user will attempt to instantiate AbstractDsxAndHitbtcFunctionality as if it was an exchange, unless we add a ton of code to workaround that (in many places, the build files, the implementation files, the transpiler, etc).

**Comment by npomfret** [comment]
> because a user will attempt to instantiate AbstractDsxAndHitbtcFunctionality

I mean, they might, but I doubt it. 

Could this class have its exchange id missing, and we blow up somewhere if someone tried to instantiate an exchange class that has no id?

**Comment by kroitor** [comment]
> Could this class have its exchange id missing, and we blow up somewhere if someone tried to instantiate an exchange class that has no id?

It would still complicate the build and we will have to amend a lot of places throughout the repository and explain the details to the users, without having the real need for the overkill and without the need for the explanations in the first place. 

In other words, to resolve this i would consider one of the two options:

- either inheritance (one exchange is tied to another)
- or just duplicate the entire integration code and rewrite certain parts of it, untying them from one another

The third option of adding special cases for abstract classes throughout the repo will be many times more complicated than any of the two options above – so, why bother?

If you find a difference in their APIs – (this is not confirmed yet) – I would simply copy-paste the entire integration code and they will not depend on each other. 

Until such a difference pops up – it works with inheritance as well. I wouldn't introduce abstract common classes to handle the difference, because it's easier and more straightforward to solve with duplication (in case such a difference exists).

**Comment by npomfret** [comment]
Sorry, I'm having a really bad day!

I've spent hours trying to get my system working but I can't figure out what is going on...  Why is the id of the BTC/USDT market _BTCUSD_ and not _BTCUSDT_?

this code:

        const exchangeId = 'dsx';
        const exchange = new ccxt[exchangeId]();
        await exchange.loadMarkets();

        console.log(exchange.markets['BTC/USDT'])

Results in:

```
{ limits:
   { amount: { min: 0.00001, max: undefined },
     price: { min: 0.01, max: undefined },
     cost: { min: 1.0000000000000001e-7, max: undefined } },
  precision: { price: 0.01, amount: 0.00001 },
  tierBased: true,
  percentage: true,
  taker: 0.0025,
  maker: 0.0015,
  info:
   { id: 'BTCUSD',
     baseCurrency: 'BTC',
     quoteCurrency: 'USD',
     quantityIncrement: '0.00001',
     tickSize: '0.01',
     takeLiquidityRate: '0.0025',
     provideLiquidityRate: '0.0015',
     feeCurrency: 'USD' },
  id: 'BTCUSD',
  symbol: 'BTC/USDT',
  base: 'BTC',
  quote: 'USDT',
  baseId: 'BTC',
  quoteId: 'USD',
  active: true,
  feeCurrency: 'USDT' }
```

And, from the cli...

`node examples/js/cli.js dsx loadMarkets --table
`
`    {"amount":{"min":0.00001},"price":{"min":0.01},"cost":{"min":1.0000000000000001e-7}} |    {"price":0.01,"amount":0.00001} |      true |       true | 0.0025 | 0.0015 |  BTCUSD | BTC/USDT |  BTC |  USDT |    BTC |     USD |   true |        USDT`

And from the exchange (https://api.dsxglobal.com/api/2/public/symbol)

```
{
"id": "BTCUSDT", <-- this is the id
"baseCurrency": "BTC",
"quoteCurrency": "USDT",
"quantityIncrement": "0.00001",
"tickSize": "0.01",
"takeLiquidityRate": "0.0025",
"provideLiquidityRate": "0.0015",
"feeCurrency": "USDT"
}
```

**Comment by npomfret** [comment]
... I don't know why, but the hitbtc version of fetch markets doesn't not work for dsx.  The BTC/USd market it being given incorrect ids:

        const exchangeId = 'dsx';
        const exchange = new ccxt[exchangeId]();
        const markets = await exchange.fetchMarkets();

        console.log(markets.find(item => item.info.id === 'BTCUSD'));

->>

```
{ tierBased: true,
  percentage: true,
  taker: 0.0025,
  maker: 0.0015,
  info:
   { id: 'BTCUSD',
     baseCurrency: 'BTC',
     quoteCurrency: 'USD',
     quantityIncrement: '0.00001',
     tickSize: '0.01',
     takeLiquidityRate: '0.0025',
     provideLiquidityRate: '0.0015',
     feeCurrency: 'USD' },
  id: 'BTCUSD',
  symbol: 'BTC/USDT', <-- wrong
  base: 'BTC',
  quote: 'USDT',  <-- wrong
  baseId: 'BTC',
  quoteId: 'USD',
  active: true,
  precision: { price: 0.01, amount: 0.00001 },
  feeCurrency: 'USDT',  <-- wrong
  limits:
   { amount: { min: 0.00001, max: undefined },
     price: { min: 0.01, max: undefined },
     cost: { min: 1.0000000000000001e-7, max: undefined } } }
```

**Comment by kroitor** [comment]
@npomfret will fix that shortly, thx for pointing it out!

**Comment by kroitor** [comment]
@npomfret fixed it in 1.28.12, thanks for pointing it out!
